Bottom: b88eede76576b8053be41f410159979f788f773b
Top:    ccb4a73ebea1fcf66d5680b28657c9a3262e2886
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-09-29 23:02:10 +0100

hid/gtk: Don't use dup_string where doing so would mix g_free() and free()

The core uses malloc(), strdup(), free() etc..
The GTK HID primarily uses g_malloc(), g_strdup(), g_free() etc..

Mixing them can lead to bad behaviour and hard to debug crashes. This
commit is not aimed at fixing any known bugs, but is "by inspection", as
I spotted some of these when working on a layer selector patch.


---

diff --git a/src/hid/gtk/gui-config.c b/src/hid/gtk/gui-config.c
index dba0635..7b2751a 100644
--- a/src/hid/gtk/gui-config.c
+++ b/src/hid/gtk/gui-config.c
@@ -169,6 +169,20 @@ static ConfigAttribute config_attributes[] = {
   {"layer-name-8", CONFIG_Unused, NULL},
 };
 
+static gboolean
+dup_core_string (gchar ** dst, const gchar * src)
+{
+  if (dst == NULL || (*dst == NULL && src == NULL))
+    return FALSE;
+
+  if (*dst != NULL && src != NULL && strcmp (*dst, src) == 0)
+    return FALSE;
+
+  free (*dst);
+  *dst = (src == NULL) ? NULL : strdup (src);
+
+  return TRUE;
+}
 
 static FILE *
 config_file_open (gchar * mode)
@@ -1362,10 +1376,10 @@ config_layers_apply (void)
     {
       layer = &PCB->Data->Layer[i];
       s = ghid_entry_get_text (layer_entry[i]);
-      if (dup_string (&layer->Name, s))
+      if (dup_core_string (&layer->Name, s))
 	layers_modified = TRUE;
 /* FIXME */
-      if (use_as_default && dup_string (&Settings.DefaultLayerName[i], s))
+      if (use_as_default && dup_core_string (&Settings.DefaultLayerName[i], s))
 	ghidgui->config_modified = TRUE;
 
     }
@@ -1421,7 +1435,7 @@ config_layers_apply (void)
   if (use_as_default)
     {
       s = make_layer_group_string (&PCB->LayerGroups);
-      if (dup_string (&Settings.Groups, s))
+      if (dup_core_string (&Settings.Groups, s))
 	{
 	  ParseGroupString (Settings.Groups, &Settings.LayerGroups, max_copper_layer);
 	  ghidgui->config_modified = TRUE;
@@ -1460,7 +1474,7 @@ layer_name_entry_cb(GtkWidget *entry, gpointer data)
 
 	layer = &PCB->Data->Layer[i];
 	name = ghid_entry_get_text(entry);
-	if (dup_string (&layer->Name, name))
+	if (dup_core_string (&layer->Name, name))
 		ghid_layer_buttons_update();
 }
 
@@ -1712,7 +1726,7 @@ config_color_defaults_cb (gpointer data)
     {
       cc = (ConfigColor *) list->data;
       ha = cc->attributes;
-      dup_string ((char **) ha->value, ha->default_val.str_value);
+      dup_core_string ((char **) ha->value, ha->default_val.str_value);
       cc->color_is_mapped = FALSE;
       ghid_set_special_colors (ha);
     }
